Ryan/feat/cnet spike trt

ryanontheinside opened the pull request:
# DRAFT 
# Add Comprehensive ControlNet Support to StreamDiffusion

## Summary

This PR adds comprehensive ControlNet support to StreamDiffusion, enabling real-time image generation conditioned on various control inputs (edges, depth, pose, etc.). The implementation provides dynamic multi-ControlNet chaining with independent conditioning strength adjustment and full TensorRT acceleration.

## Key Features

**Dynamic Multi-ControlNet Support**
- Chain multiple ControlNets for complex conditioning (depth + canny + pose, etc.)
- Independent conditioning strength adjustment at runtime
- Any combination of ControlNets work together seamlessly

**Full TensorRT Acceleration**
- Automated TensorRT engine building and management
- Both UNet and ControlNet models accelerated
- 2x speedup on ControlNet inference (~7ms savings per frame on 5090)

**Graceful Degredation**
- Falls back to pytorch inference should tensorrt acceleration be unavailable for controlnets

**Performance (depth controlnet, tensorrt depth anything v2)**
- 27.8-28 fps on windows with a 4090, 65% GPU usage
  - driver 572.83 and cuda 12.8, cpu - amd ryzen 5 7600X (6 cores/12 threads), ddr5 ram
- 38-40 fps on linux with a 4090, 75% GPU usage (this is same one I have on staging for live AI video)
  - driver 570.133.07, cuda 12.8, cpu - intel i5 12600KF (16 vcores - 6 performance, 4 efficient), ddr4 ram
- 44-45 fps on 5090 in linux, 65% GPU usage
  - driver 570.144, cuda 12.8 driver cpu - threadripper pro 7965WX, ddr5 ram

**Developer API**
- Drop-in replacement for standard StreamDiffusion
- YAML/JSON configuration files for easy pipeline setup
- Extensible preprocessor system for custom ControlNets
- Runtime parameter adjustment via GUI and programmatic API

## Architecture Support
- ✅ **SD 1.5**: Full support with extensive ControlNet ecosystem
- ✅ **SD 2.1**: Full support with TensorRT acceleration
- 🚧 **SDXL**: Forthcoming (broken during refactor, being restored)

## Supported ControlNet Types

Preprocessors are the limiting factor. New preprocessors are being added often.

## Basic Usage

**For comprehensive usage examples, configuration formats, and API documentation, see `README_CONTROLNET.md`**

## Implementation Details

### TensorRT Integration
- **Automated engine management**: Checks for existing engines, builds if needed
- **Dynamic batch sizing**: Handles variable frame buffer sizes efficiently  
- **Memory optimization**: Mild buffer allocation strategies (optimization to base StreamDiffusion)
- **Fallback support**: Graceful degradation to PyTorch if TensorRT unavailable via `HybridControlNet` class

### Multi-ControlNet Architecture
- **Independent preprocessing pipelines**: Each ControlNet runs its own preprocessor, with simple parallelism 
- **Efficient conditioning**: Optimized tensor operations for multiple control inputs
- **Runtime flexibility**: Add/remove/adjust ControlNets without pipeline recreation
- **Hybrid acceleration**: Each ControlNet automatically uses TensorRT if available, falls back to PyTorch

## Demo Applications

### Web demo
```bash
cd demo/realtime-img2img
python main.py
```
Select a YAML configuration file, and click start.  The demo is built upon the existing streamdiffusion demo, but now includes:
- Real-time parameter adjustment
- Multi-ControlNet strength sliders
- Live performance monitoring
- Side-by-side input/output display

### High-Performance Standalone Demo
```bash
python examples/controlnet/standalone_controlnet_pipeline.py
```
- Production-ready reference implementation
- Minimal overhead for maximum performance
- Batch processing capabilities
- Export functionality

### Benchmarking demo

You may use this script to benchmark configurations. It accepts a video as input. Run it with a command similar to this one:

```bash
python examples\controlnet\controlnet_video_test.py --config .\configs\controlnet_examples\sdturbo_depth_trt_example.yaml --input-video ..\input\xilin_talk_video.mp4
```

## File Structure

```
src/streamdiffusion/controlnet/
├── __init__.py                           # Public API exports
├── base_controlnet_pipeline.py          # Core ControlNet pipeline logic (1062 lines)
├── controlnet_pipeline.py               # SD 1.5/2.1 implementation  
├── controlnet_sdxlturbo_pipeline.py     # SDXL implementation (WIP)
├── config.py                            # Configuration management
└── preprocessors/                       # Preprocessor implementations
    ├── __init__.py                      # Preprocessor registry
    ├── base.py                          # Base preprocessor class
    ├── canny.py                         # Canny edge detection
    ├── depth.py                         # MiDaS depth estimation
    ├── depth_tensorrt.py                # TensorRT depth (DepthAnything)
    ├── openpose.py                      # Human pose detection
    ├── lineart.py                       # Line art extraction
    └── passthrough.py                   # No preprocessing

src/streamdiffusion/acceleration/tensorrt/
├── __init__.py                          # TensorRT acceleration exports
├── controlnet_engine.py                # HybridControlNet class & engine management
├── controlnet_models.py                # ControlNet TensorRT model definitions
├── controlnet_wrapper.py               # ControlNet TensorRT wrapper
├── engine_pool.py                      # Engine caching and compilation
├── models.py                           # Base TensorRT model implementations
├── utilities.py                        # TensorRT utilities
├── model_detection.py                  # Model architecture detection
├── engine.py                           # Base engine management
└── builder.py                          # TensorRT engine building

examples/controlnet/
├── controlnet_webcam_gui_demo.py        # Interactive GUI demo
└── standalone_controlnet_pipeline.py    # Production reference

configs/controlnet_examples/
├── canny_example.yaml                   # Basic canny configuration
├── depth_example.yaml                   # MiDaS depth setup
├── depth_trt_example.yaml              # TensorRT depth setup
├── lineart_example.yaml                # Line art configuration
├── multi_controlnet_example.yaml        # Multi-ControlNet demo
├── sdturbo_canny_example.yaml           # SD-Turbo canny setup
├── sdturbo_depth_trt_example.yaml       # SD-Turbo TensorRT depth
├── sdturbo_depth_trt_example_4steps.yaml # 4-step variant
├── sdxlturbo_canny_example.yaml         # SDXL canny (WIP)
├── sdxlturbo_depth_trt_example.yaml     # SDXL depth (WIP)
└── tile_example.yaml                    # Tile ControlNet setup
```

## Technical Achievements

### TensorRT Optimization Pipeline  
- Dynamic multi-controlnet support
- Controlnet compatability with TensorRT Unet
- Controlnet acceleration with TensorRT
- Automated TensorRT engine building with dependency management
- Dynamic engine caching and version management
- Optimized memory allocation patterns
- Support for multiple ControlNet engine compilation

### Runtime Flexibility
- Dynamic ControlNet strength adjustment without pipeline recreation
- Efficient control image preprocessing pipeline
- Multi-ControlNet coordination with independent control
- Memory-efficient batch processing for real-time applications

## Testing & Validation

### Performance Testing
- Benchmarked on RTX 5090 with various configurations
- Profiling-driven optimization validation  
- Regression testing for performance improvements
- Memory usage optimization verification

### Functional Testing
- Multi-ControlNet combinations tested
- Configuration file validation
- TensorRT engine building automation
- Runtime parameter adjustment
- Error handling and graceful degradation

### Compatibility Testing
- SD 1.5 model compatibility
- SD 2.1 and SD-Turbo model support
- Various ControlNet model architectures
- Different TensorRT versions and CUDA environments

## Breaking Changes

None. This implementation is designed as a drop-in addition to StreamDiffusion with full backward compatibility.

## Future Work & Roadmap

### Immediate (Next Sprint)
- **Restore SDXL support** - Fix SDXL pipeline broken during refactor
- **Make config system agnostic to controlnet** - The system should handle configs without controlnets. 
- **Frame_buffer_size issue** - Solve issue of tensorrt acceleration of Unet with frame_buffer_size > 1


### Short Term 
- **Expanded preprocessor optimizations** - Bring other preprocessors up[ to depth_trt standard
- **Additional preprocessors** - HED, normal maps, semantic segmentation
- **Further TensorRT optimizations** - More aggressive buffer allocation
- **Pipeline caching** - Reduce initialization overhead for repeated usage

### Medium Term 
- **multi stage processing** - Add multistage processing to system

ryanontheinside committed:
initial webcam test

ryanontheinside committed:
spike

ryanontheinside committed:
initial test with controlnet

ryanontheinside committed:
lineart script, controlnet script test

ryanontheinside committed:
general purpose single controlnet pipe, addnl configs

ryanontheinside committed:
restore demo script, readme

ryanontheinside committed:
fix controlnet preprocess, update depth example

ryanontheinside committed:
add some models to readme for referece

ryanontheinside committed:
update default lineart controlnet

ryanontheinside committed:
add tile

ryanontheinside committed:
multi demo

ryanontheinside committed:
move examples

ryanontheinside committed:
remove old script

ryanontheinside committed:
preprocess optim

ryanontheinside committed:
gpu prepro

ryanontheinside committed:
sdturbo and sdxl controlnet

# Conflicts:
#	examples/controlnet/controlnet_webcam_demo.py

ryanontheinside committed:
sdturbo streamdiffusion controlnet

ryanontheinside committed:
SDXLTurbo controlnet streamdiffusion

ryanontheinside committed:
monolithic pipeline refactor

ryanontheinside committed:
high performance controlnet adaptation demo

ryanontheinside committed:
consolidate sd1.5 + sdturbo

ryanontheinside committed:
update examples

ryanontheinside committed:
replace strength w t_index_list

ryanontheinside committed:
remove unneeded branching

ryanontheinside committed:
ignore onnx

ryanontheinside committed:
sdxl tensor fix

ryanontheinside committed:
refactor - sdxl broken

this is to refactor to follow the existing codebase architecture more closely.
with a tight timeline, SDXL support is not a priority. TODO
todo: look at state before this commit and fix SDXL

ryanontheinside committed:
spike

ryanontheinside committed:
model detection update

ryanontheinside committed:
reference documents

ryanontheinside committed:
remove duplicate

ryanontheinside committed:
trt compiling

ryanontheinside committed:
yaml update

ryanontheinside committed:
dynamic controlnet + dynamic conditioning streen

ryanontheinside committed:
working example multicontrolnet

ryanontheinside committed:
remove extranneous scrtipt

use the regular webcam demo

ryanontheinside committed:
varying spatial dimensions

ryanontheinside committed:
remove print statements

ryanontheinside committed:
remove profiling

ryanontheinside committed:
image conversions

ryanontheinside committed:
cached preprocessor

ryanontheinside committed:
fix overwriting config

ryanontheinside committed:
phase 2 plan

ryanontheinside committed:
tensorrt controlnet

ryanontheinside committed:
remove debug scripts

ryanontheinside committed:
restore test script

ryanontheinside committed:
remove async compialtion

ryanontheinside committed:
batch dimension

ryanontheinside committed:
batch size detection for controlnets

ryanontheinside committed:
partial cnet trt - black frames

ryanontheinside committed:
working controlnet trt

ryanontheinside committed:
delete test

ryanontheinside committed:
webcam gui

ryanontheinside committed:
gui improvement

ryanontheinside committed:
frame buffer size, delta support

ryanontheinside committed:
frame buffer delta demo

ryanontheinside committed:
restrict frame buffer size

ryanontheinside committed:
notes

ryanontheinside committed:
demo frame buffer size

ryanontheinside committed:
clarity

ryanontheinside committed:
config use denoising batch

ryanontheinside committed:
remvoe unneeded tensor cloning in controlnet engin

ryanontheinside committed:
optimizations

ryanontheinside committed:
optimization

ryanontheinside committed:
optimize postprocess pipeline

ryanontheinside committed:
tensorrt buffer allocation

ryanontheinside committed:
get_controlnet_conditioning cleanup

ryanontheinside committed:
remove extranneous scripts

ryanontheinside committed:
update readme

ryanontheinside committed:
update multi controlnet example

ryanontheinside committed:
stand alone simple pipeline

ryanontheinside committed:
remove controlnet config abstraction

removes nonessential dataclass abstraction. Streamdiffusion uses dicts

ryanontheinside committed:
remove demo function

ryanontheinside committed:
remove legacy demo

ryanontheinside committed:
update docstrings

ryanontheinside committed:
update docstring

ryanontheinside committed:
update readme

ryanontheinside committed:
stack trace in controlnet fallback

ryanontheinside committed:
add yolonas pose tensorrt preprocessor

ryanontheinside committed:
attribution

ryanontheinside committed:
demo exit behavior

ryanontheinside committed:
selectable config

ryanontheinside committed:
video test

ryanontheinside committed:
mediapipe preprocessor spike

ryanontheinside committed:
touch

ryanontheinside committed:
add engine only mode

ryanontheinside committed:
todo

ryanontheinside committed:
mediapipe segementation preprocessor

ryanontheinside committed:
theoretical multi-stage processing plan

ryanontheinside committed:
tensorrt buffer allocation fix

ryanontheinside committed:
controlnet support in web demo, partial

tunuous support for default pipeline

ryanontheinside committed:
standard lineart preprocessor spike

ryanontheinside committed:
standar lineart optimization

ryanontheinside committed:
simple preprocessing parallelism

ryanontheinside committed:
live prompt update

ryanontheinside committed:
realtime tindex list update

ryanontheinside committed:
add tindex real time update

ryanontheinside committed:
update interface, tindex list

ryanontheinside committed:
respect acceleration

ryanontheinside committed:
remove amorphous yamls

ryanontheinside committed:
clean up examples and readme

ryanontheinside committed:
consistent naming convention

ryanontheinside committed:
config system controlnet agnostic

ryanontheinside committed:
add additional realtime params

ryanontheinside committed:
remove notes

ryanontheinside committed:
respect engine dir in wrapper

ryanontheinside committed:
respect engine dir in demo

ryanontheinside committed:
remove antiquated control methods

ryanontheinside committed:
demo header

ryanontheinside committed:
parameter update documentation

ryanontheinside committed:
TODO

ryanontheinside committed:
remove antiquated method from test script

ryanontheinside committed:
add optional config to standalone pipeline demo

ryanontheinside committed:
remove gui demo

ryanontheinside committed:
use stream resolution

ryanontheinside committed:
use utility method to load config

ryanontheinside committed:
respect engine dir

eliteprox commented on src/streamdiffusion/acceleration/tensorrt/controlnet_models.py:
```diff
@@ -0,0 +1,207 @@
+                 controlnet_type: str = "canny",
+                 fp16: bool = True,
+                 device: str = "cuda",
+                 max_batch: int = 2,
+                 min_batch_size: int = 1,
```
@ryanontheinside Can this be named `max_batch_size` or rename `min_batch_size` to `min_batch` to keep the param names consistent? Curious what led to this change, is there a conflict with another param downstream?

ai-runner change was needed https://github.com/livepeer/ai-runner/pull/646/commits/cc9d103a71844135d18c29472275a14d0d7ce130

ryanontheinside committed:
fix variable

ryanontheinside committed:
duplicate cfg logic, comment reduction

ryanontheinside committed:
remove duplicative update_controlnet_scale

this is in wrapper

ryanontheinside committed:
remove duplicative get_last_processed_image

this is in wrapper.py

ryanontheinside committed:
remove two dead methods

ryanontheinside committed:
docstring & comment reduction

ryanontheinside committed:
remove get input output names

dead methods

ryanontheinside committed:
docstring fluff and comment reduction

ryanontheinside committed:
remove dead methods

ryanontheinside committed:
docstring and comment reduction

ryanontheinside committed:
docstring & comment reduction

ryanontheinside committed:
remove deprecated method

ryanontheinside committed:
remove deprecated save_contronet_config

ryanontheinside committed:
remove unused method

ryanontheinside committed:
Revert "remove dead methods"

This reverts commit 9913223929236ac38fd579b124724830b0e85942.

ryanontheinside committed:
docstring & comment reduction

ryanontheinside committed:
restore update controlnet scale

ryanontheinside committed:
cleanp docstring and comments

ryanontheinside committed:
docstring & comment reduction

ryanontheinside committed:
docstring comment reduction

ryanontheinside committed:
comment cleanup

ryanontheinside committed:
readme update

ryanontheinside committed:
readme file name

ryanontheinside committed:
update stand alone example

victorges committed:
Copy changes from ai-runner wrapper

victorges committed:
Remove copied header

We're original now

victorges committed:
Move wrapper to main streamdiffusion package

ryanontheinside committed:
obey config resolution

ryanontheinside committed:
move config

ryanontheinside committed:
use direct instead of refelction

victorges committed:
wrapper: Merge with ai-runner and export from lib (#3)

* Copy changes from ai-runner wrapper

* Remove copied header

We're original now

* Export StreamDiffusionWrapper for ai-runner

So we can avoid copying the whole thing over there.

* Move wrapper to main streamdiffusion package

* Fix rebase typo

* obey config resolution

* unet multires trt

* controlnet engine naming convention

* preprocessor refactor

This commit reduces code duplication without changing the public interface or performance.
Next steps: more work on consistent GPU first approach
BEFORE REFACTOR
--------------------------------------------------------------------------------
Preprocessor              Avg Time (s) Min (s)  Max (s)  Std (s)  Output Size
--------------------------------------------------------------------------------
Passthrough_PIL           0.000        0.000    0.000    0.000    (512, 512)
Passthrough_Tensor        0.001        0.000    0.001    0.000    (512, 512)
Canny_PIL                 0.001        0.001    0.002    0.000    (512, 512)
Canny_Tensor              0.001        0.001    0.002    0.000    (512, 512)
StandardLineart_PIL       0.003        0.002    0.004    0.001    (512, 512)
StandardLineart_Tensor    0.004        0.003    0.004    0.001    (512, 512)
Depth_PIL                 0.024        0.021    0.028    0.002    (512, 512)
Depth_Tensor              0.024        0.022    0.025    0.001    (512, 512)
Lineart_PIL               0.157        0.155    0.159    0.002    (512, 512)
Lineart_Tensor            0.164        0.158    0.169    0.004    (512, 512)
OpenPose_PIL              0.362        0.351    0.368    0.007    (512, 512)
OpenPose_Tensor           0.361        0.348    0.371    0.009    (512, 512)
MediaPipePose_PIL         0.024        0.024    0.026    0.001    (512, 512)
MediaPipePose_Tensor      0.027        0.026    0.029    0.001    (512, 512)
MediaPipeSegmentation_PIL 0.004        0.003    0.004    0.001    (512, 512)
MediaPipeSegmentation_Tensor 0.006        0.005    0.006    0.000    (512, 512)
PoseTensorRT_PIL          0.021        0.020    0.022    0.001    (512, 512)
PoseTensorRT_Tensor       0.015        0.015    0.016    0.000    (512, 512)
DepthTensorRT_PIL         0.005        0.004    0.006    0.001    (512, 512)
DepthTensorRT_Tensor      0.002        0.002    0.002    0.000    (512, 512)

AFTER CHANGES
Preprocessor              Avg Time (s) Min (s)  Max (s)  Std (s)  Output Size
--------------------------------------------------------------------------------
Passthrough_PIL           0.000        0.000    0.000    0.000    (512, 512)
Passthrough_Tensor        0.000        0.000    0.000    0.000    (512, 512)
Canny_PIL                 0.001        0.000    0.001    0.000    (512, 512)
Canny_Tensor              0.001        0.001    0.002    0.000    (512, 512)
StandardLineart_PIL       0.003        0.002    0.004    0.001    (512, 512)
StandardLineart_Tensor    0.004        0.004    0.005    0.001    (512, 512)
Depth_PIL                 0.024        0.022    0.026    0.002    (512, 512)
Depth_Tensor              0.024        0.023    0.025    0.001    (512, 512)
Lineart_PIL               0.160        0.158    0.164    0.003    (512, 512)
Lineart_Tensor            0.160        0.155    0.164    0.003    (512, 512)
OpenPose_PIL              0.355        0.352    0.358    0.002    (512, 512)
OpenPose_Tensor           0.367        0.346    0.383    0.014    (512, 512)
MediaPipePose_PIL         0.024        0.023    0.025    0.001    (512, 512)
MediaPipePose_Tensor      0.027        0.026    0.027    0.000    (512, 512)
MediaPipeSegmentation_PIL 0.003        0.003    0.004    0.000    (512, 512)
MediaPipeSegmentation_Tensor 0.005        0.005    0.006    0.000    (512, 512)
PoseTensorRT_PIL          0.021        0.020    0.023    0.001    (512, 512)
PoseTensorRT_Tensor       0.016        0.014    0.018    0.002    (512, 512)
DepthTensorRT_PIL         0.004        0.004    0.005    0.001    (512, 512)
DepthTensorRT_Tensor      0.002        0.002    0.004    0.001    (512, 512)

* external preprocessor

* move config

* update img2img demo import

* seed update

* provide dummy inputs

---------

Co-authored-by: RyanOnTheInside <7623207+ryanontheinside@users.noreply.github.com>

ryanontheinside committed:
prompt blending working

ryanontheinside committed:
seed blending

ryanontheinside committed:
prompt and seed blending

ryanontheinside committed:
interface update

ryanontheinside committed:
support text to image configs

ryanontheinside committed:
fix dupes from rebase

ryanontheinside committed:
add optional weighting for prompt and seed

ryanontheinside committed:
remove pipeline type and clean up model loading

ryanontheinside committed:
remove infer_model_type

ryanontheinside committed:
remove infer_controlnet_type

ryanontheinside committed:
remove pipeline_type from example configs

ryanontheinside committed:
remove ppipeline_type from demo

ryanontheinside committed:
documentation

ryanontheinside committed:
remove pipeline from config

ryanontheinside committed:
remove pipline_type from demo

ryanontheinside committed:
remove resolution

ryanontheinside committed:
Merge pull request #8 from livepeer/ryan/feat/controlnet-video-test-cleanup

Ryan/feat/controlnet video test cleanup

ryanontheinside committed:
remove .lower

ryanontheinside committed:
happy path

ryanontheinside committed:
Merge pull request #7 from livepeer/ryan/feat/model-detection

Ryan/feat/model detection

ryanontheinside committed:
soft edge preprocessor

ryanontheinside committed:
HED preprocessor

ryanontheinside committed:
remove custom network

ryanontheinside committed:
pipelined processing

this results in 20%-30% fps improvement when using four controlnets

ryanontheinside committed:
update public interface and use it

ryanontheinside committed:
remove overly defensive checks

ryanontheinside committed:
complete parameter updater forwarding methods

ryanontheinside committed:
Merge pull request #11 from livepeer/ryan/feat/option-weighting-update

Ryan/feat/option weighting update

ryanontheinside committed:
Merge pull request #6 from livepeer/ryan/feat/option-weighting

Optional prompt and seed weight normalization

ryanontheinside committed:
Merge pull request #9 from livepeer/ryan/feat/soft-edge-preprocessor

Ryan/feat/soft edge preprocessor

ryanontheinside committed:
update dynamic controls

ryanontheinside committed:
cleanup prints

ryanontheinside committed:
vectorize middle joints

ryanontheinside committed:
precompute edges and colors

ryanontheinside committed:
reduce cuda available system calls

ryanontheinside committed:
prediction list and tuple creation

ryanontheinside committed:
Merge pull request #13 from livepeer/ryan/feat/pose_tensorrt-preprocessor-improvements

Ryan/feat/pose tensorrt preprocessor improvements

ryanontheinside committed:
Merge pull request #12 from livepeer/ryan/feat/prompt-weighting-demo

update dynamic controls

ryanontheinside committed:
reduce detection resolution

ryanontheinside committed:
vectorize

ryanontheinside committed:
video mode, gpu delegate

ryanontheinside committed:
Merge pull request #14 from livepeer/ryan/feat/mediapipe-pose-improvements

Ryan/feat/mediapipe pose improvements

ryanontheinside committed:
orchestration module

ryanontheinside committed:
Merge pull request #10 from livepeer/ryan/feat/pipelined-preprocessing

Ryan/feat/pipelined preprocessing

ryanontheinside committed:
add normalize consistency

ryanontheinside committed:
generic cache stats and normalization

ryanontheinside committed:
unified blending control

ryanontheinside committed:
rename interpolation method to prompt interpolation method

ryanontheinside committed:
Merge pull request #19 from livepeer/ryan/fix/normalization-behavior

Ryan/fix/normalization behavior

ryanontheinside committed:
forward legacy single prompt

ryanontheinside committed:
Merge pull request #20 from livepeer/ryan/fix/forward-legacy-prompt

forward legacy single prompt

ryanontheinside committed:
remove legacy prompt box

# Conflicts:
#	demo/realtime-img2img/frontend/src/lib/components/PromptBlendingControl.svelte
#	demo/realtime-img2img/frontend/src/routes/+page.svelte

ryanontheinside committed:
fix prompt blending

# Conflicts:
#	demo/realtime-img2img/frontend/src/lib/components/SeedBlendingControl.svelte
#	demo/realtime-img2img/main.py

ryanontheinside committed:
unified blending control and fixes

# Conflicts:
#	demo/realtime-img2img/frontend/src/lib/components/BlendingControl.svelte

victorges committed:
Set "normalize weight" params w/ update params func (#24)

BuffMcBigHuge committed:
Multi-Resolution (Pipeline Reload) + WebUI Upgrades (#15)

* Added api-only mode to server, update svelte vite and frontend packages, built multiresolution support with tensorrt, built resolution picker.

* Added realtime resolution updates.

* Small fix.

* Controlnet work.

* ControlNet testing.

* Rebuilt resolution swapping by restarting pipeline, re-added prompt input to UI for testing.

* Fix missing file.

* Fix to config updating the resolution.

* fix input profiles

* restore sdxl functionality

* remove legacy prompt box

* fix prompt blending

* unified blending control and fixes

* Set dynamic size from 384-1024 with opt at 704.

* updater: Remove "dynamic" resolution update logic

* Remove debug logs

* Remove more debug logs...

---------

Co-authored-by: RyanOnTheInside <7623207+ryanontheinside@users.noreply.github.com>
Co-authored-by: Victor Elias <victorgelias@gmail.com>
Co-authored-by: Victor Elias <victor@livepeer.org>

ryanontheinside committed:
configurable pytorch fallback

ryanontheinside committed:
rename interpolation method

ryanontheinside committed:
remove legacy files

ryanontheinside committed:
Merge pull request #26 from livepeer/ryan/fix/pytorch-fallback

configurable pytorch fallback

ryanontheinside committed:
Merge pull request #27 from livepeer/ryan/fix/demo-prompt-fix

Ryan/fix/demo prompt fix

victorges committed:
IPAdapters + Controlnet and Tensorrt for SDXL (#40)

* testing

* text to image working

* img2img working

* example update

* update examples

* successful engine build

* ipadapter tensorrt acceleration

* remove test scripts

* remove test script

* safe import

* update test script

* working

* button up

* bake with strength | name engine dir

* cleanup wrapper

* cache embeddings

* comment some prints

* backtrack multiple ipadapter support

this will be done in subsequent phase

* remove test scripts

* dynamic ipadapter token

* plus support

* rm erroneous commit

* move config

* erroneous comment

* demo spike

* prompts working

* composable conditioning wrapper

* bugfix wrapper

* wrapper fix

* remove print statement

* fix prompt embeds

* add ipadapter support to demo

* remove extranneous exports

* remove extranneous backwards compat

* remove extrannous controlnet exports

* comment cleanup

* rm ipadapter readme

* cleanup prints

* remove debug script

* cleanup config logic

* remove defensive import handling

* remove example bloat

* remove extranneous demo

* rm print statements

* remove prints, update todo

* fix config upload

* fix  style image update

* remove copywritten image

* leverage preprocessor system

* remove print statements

* prompt update perf fix

* remove heuristic ipadapter detection

# Conflicts:
#	src/streamdiffusion/ipadapter/base_ipadapter_pipeline.py

* Ryan/feat/02 ipadapter on sdxl (#32)

* merge sdxl

Merged changes from ryan/feat/ipadapter-on-sdxl branch with marco/feat/sdxl branch, resolving conflicts in:
src/streamdiffusion/pipeline.py: Integrated SDXL UNet calling conventions with comprehensive conditioning support, retained SDXL added_cond_kwargs handling and TensorRT vs PyTorch UNet detection
demo/realtime-img2img/main.py: Combined IPAdapter info response with resolution calculation, merged pipeline creation logic for both IPAdapter and SDXL support
src/streamdiffusion/acceleration/tensorrt/__init__.py: Unified embedding dimension detection for both SDXL (2048) and SD1.5 (768), integrated IPAdapter and ControlNet TensorRT support
src/streamdiffusion/controlnet/base_controlnet_pipeline.py: Merged ControlNet conditioning logic with SDXL compatibility
src/streamdiffusion/stream_parameter_updater.py: Integrated IPAdapter embedding enhancers with parameter update system
src/streamdiffusion/wrapper.py: Resolved TensorRT engine compilation conflicts, unified IPAdapter and SDXL model loading paths

* handle sdxl in conditioning wrapper

* image encoder download fix

* remove extranneous ipadapter pipeline abstraction

* remove heuristic ipadapter detection

* consolidate model detection

* Ryan/feat/03 controlnet on ipadapter on sdxl (#33)

* controlnet enabled unet engine

* remove prints

* cnet engine support

* consolidate utilities logging

* uncomment fallback

* consolidate sdxl unet wrapper

* Fixed model detection.

* Rebuilt model detection mechanism, cleanup.

* Removal of fluff.

* More reliance on detect_model function, fix of SDXL detection for onnx export.

* Bit of cleanup and additional logging.

* Reverted controlnet/ipadapter model detection to original functions.

* add controlnet model detection

* add controlnet model detection to export

---------

Co-authored-by: BuffMcBigHuge <marco@bymar.co>

---------

Co-authored-by: BuffMcBigHuge <marco@bymar.co>

---------

Co-authored-by: RyanOnTheInside <7623207+ryanontheinside@users.noreply.github.com>
Co-authored-by: BuffMcBigHuge <marco@bymar.co>

ryanontheinside committed:
rename to export wrapper convention

ryanontheinside committed:
move sdxl controlnet export

ryanontheinside committed:
rename conditioning wrapper

ryanontheinside committed:
rename controlnet_wrapper.py

ryanontheinside committed:
rename ipadapter_wrapper.py

ryanontheinside committed:
rename sdxl_support.pt, move SDXLExportWrapper

ryanontheinside committed:
move export wrappers to directory

ryanontheinside committed:
move models

ryanontheinside committed:
rename engine.py

ryanontheinside committed:
move engines to directory

ryanontheinside committed:
fix straggle imports

ryanontheinside committed:
move preprocessors from src/streamdiffusion/controlnet/ --> src/streamdiffusion

ryanontheinside committed:
move engine pool from runtime folder

ryanontheinside committed:
install and use diffusers_ipadapter package

ryanontheinside committed:
Merge pull request #43 from livepeer/ryan/refactor/03-ipadapter-as-package

ryan/refactor/03-ipadapter-as-package

ryanontheinside committed:
Merge pull request #42 from livepeer/ryan/refactor/02-preprocessor

Ryan/refactor/02 preprocessor

ryanontheinside committed:
Merge pull request #41 from livepeer/ryan/refactor/01-trt-wrapper-conventions

Ryan/refactor/01 trt wrapper conventions, ipadapter as package, preprocessor refactor

ryanontheinside committed:
remove legacy docs

ryanontheinside committed:
Merge pull request #44 from livepeer/ryan/fix/remove-legacy-docs

remove legacy docs

ryanontheinside committed:
set use cuda graph true in wrapper

ryanontheinside committed:
Merge pull request #45 from livepeer/ryan/fix/use-cuda-graph

set use cuda graph true in wrapper

ryanontheinside committed:
Ryan/fix/01 fix perf regression (#46)

* remove logging, input/output validation

* remove logging, inputoutput from infer()

* remove logging and input output validation from unet

* remove overhead from hotpath

remove logging, input/output validation, make vram montioring in debug mode only

* remove logging from controlnet engine

* remove logger.debug

* remove HybridControlnet abstraction and pytorch fallback

* remove dead code

* add feedback preprocessor

* add feedback preprocessor support

* add blend to preprocessor

* add preprocessor metadata

* dynamic preprocessor controls

* remove erroneous variable

* unify tensorrt engine management

* remove unicode chars

* import fix

* change default acceleration

* fullscreen

* add route to get default image and makes ipadapter use it

* rgb conversion fix

* static video input control

* minimalist resolution picker

* increase max delta value

* Update resolution button
